File: frontend/src/api/repositorio.ts
import { apiFetch } from './client'

/** Orígenes de hojas de vida soportados por el backend (HU-05). */
export type TipoRepositorio = 'ONEDRIVE' | 'GDRIVE'

/**
 * Credenciales del origen. La forma depende del `tipo`:
 *
 * - ONEDRIVE: tenant_id, client_id, client_secret, drive_id
 * - GDRIVE:   credenciales_json (el archivo de la cuenta de servicio)
 *
 * Los campos secretos pueden omitirse: el backend reutiliza el valor
 * almacenado del mismo proveedor.
 */
export type Credenciales = Record<string, string>

export interface ConexionRepositorio {
  configurado: boolean
  tipo: TipoRepositorio | null
  carpeta: string | null
  /** El backend nunca devuelve los secretos en claro, solo enmascarados. */
  credenciales: Credenciales
  secreto_legible: boolean
  ultima_validacion: string | null
  documentos_detectados: number | null
}

export interface ConexionPayload {
  tipo: TipoRepositorio
  carpeta: string
  credenciales: Credenciales
}

export interface PruebaConexion {
  exito: boolean
  documentos_detectados: number
  mensaje: string
  detalle: string | null
}

export interface GuardarConexion {
  guardado: boolean
  prueba: PruebaConexion
  configuracion: ConexionRepositorio | null
}

export async function obtenerConexion(): Promise<ConexionRepositorio> {
  return apiFetch<ConexionRepositorio>('/configuracion/repositorio')
}

/** Valida sin guardar. */
export async function probarConexion(datos: ConexionPayload): Promise<PruebaConexion> {
  return apiFetch<PruebaConexion>('/configuracion/repositorio/probar', {
    method: 'POST',
    body: JSON.stringify(datos),
  })
}

/** Revalida la conexión ya guardada y actualiza el conteo de documentos. */
export async function probarConexionGuardada(): Promise<PruebaConexion> {
  return apiFetch<PruebaConexion>('/configuracion/repositorio/probar-guardada', { method: 'POST' })
}

/** Guarda solo si la conexión es válida; si no, la anterior queda intacta. */
export async function guardarConexion(datos: ConexionPayload): Promise<GuardarConexion> {
  return apiFetch<GuardarConexion>('/configuracion/repositorio', {
    method: 'PUT',
    body: JSON.stringify(datos),
  })
}

File: backend/schemas/repositorio.py
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field

from backend.schemas.onedrive import PruebaConexionResponse  # noqa: F401  (reexport)

TipoRepositorio = Literal["ONEDRIVE", "GDRIVE"]


class ConexionRepositorioRequest(BaseModel):
    """
    Datos de conexión al origen de hojas de vida (HU-05), sea cual sea el
    proveedor.

    `credenciales` es un diccionario cuya forma depende de `tipo`:

    - ONEDRIVE: tenant_id, client_id, client_secret, drive_id
    - GDRIVE:   credenciales_json (el archivo de la cuenta de servicio)

    Los campos secretos pueden omitirse o enviarse vacíos: se reutiliza el
    valor almacenado del mismo proveedor. Así se corrige la carpeta sin volver
    a escribir la credencial.
    """

    tipo: TipoRepositorio
    carpeta: str = Field(default="", max_length=400)
    credenciales: dict[str, str] = Field(default_factory=dict)


class ConexionRepositorioResponse(BaseModel):
    """
    Configuración actual. Los secretos siempre viajan enmascarados (RNF10): la
    API nunca devuelve la credencial en claro.
    """

    configurado: bool = False
    tipo: Optional[TipoRepositorio] = None
    carpeta: Optional[str] = None
    credenciales: dict[str, str] = Field(default_factory=dict)
    secreto_legible: bool = True
    ultima_validacion: Optional[datetime] = None
    documentos_detectados: Optional[int] = None


class GuardarRepositorioResponse(BaseModel):
    guardado: bool
    prueba: PruebaConexionResponse
    configuracion: Optional[ConexionRepositorioResponse] = None
